src/esp/geo/VoxelGrid.cpp

```diff
@@ -0,0 +1,323 @@
+  }
+
+  for (const auto& norm : coneNormals) {
+    normals.push_back(norm);
+  }
+
+  for (const auto&& index : coneIndices) {
+    indices.push_back(sz + index[0]);
+  }
+
+  // render cylinder (arrow stem)
+  Mn::Trade::MeshData cylinderData = Mn::Primitives::cylinderSolid(1, 3, 1.0f);
+
+  sz = positions.size();
+  const auto&& cylinderPositions = cylinderData.positions3DAsArray();
+  const auto&& cylinderNormals = cylinderData.normalsAsArray();
+  const auto&& cylinderIndices = cylinderData.indices();
+
+  for (const auto& ind : cylinderPositions) {
+    positions.push_back(
+        vecRotation.transformVector(ind * Mn::Vector3(0.01, 0.03, 0.01) -
+                                    Mn::Vector3(0, 0.025, 0)) +
+        mid);
+    colors.push_back(Mn::Color3(0, .3, 1));
+  }
+
+  for (const auto& norm : cylinderNormals) {
+    Mn::Vector3 transformedNorm =
+        vecRotation.transformVector(norm * Mn::Vector3(0.01, 0.04, 0.01));
+    normals.push_back(transformedNorm);
+  }
+
+  for (const auto&& index : cylinderIndices) {
+    indices.push_back(sz + index[0]);
+  }
+}
+
+void VoxelGrid::generateMeshDataAndMeshGL(const std::string& gridName,
+                                          std::vector<Mn::UnsignedInt>& indices,
+                                          std::vector<Mn::Vector3>& positions,
+                                          std::vector<Mn::Vector3>& normals,
+                                          std::vector<Mn::Color3>& colors) {
+  // If the mesh already exists, replace it. Otherwise, create a new entry in
+  // the dict
+  Mn::Trade::MeshData positionMeshData{
+      Mn::MeshPrimitive::Triangles,
+      {},
+      indices,
+      Mn::Trade::MeshIndexData{indices},
+      {},
+      positions,
+      {Mn::Trade::MeshAttributeData{Mn::Trade::MeshAttribute::Position,
+                                    Cr::Containers::arrayView(positions)}}};
+  meshDataDict_[gridName] =
+      std::make_shared<Mn::Trade::MeshData>(Mn::MeshTools::interleave(
+          positionMeshData,
+          {Mn::Trade::MeshAttributeData{Mn::Trade::MeshAttribute::Color,
+                                        Cr::Containers::arrayView(colors)},
+           Mn::Trade::MeshAttributeData{Mn::Trade::MeshAttribute::Normal,
+                                        Cr::Containers::arrayView(normals)}}));
+
+  // If the mesh already exists, replace it. Otherwise, create a new entry in
+  // the dict
+  if (meshGLDict_.find(gridName) != meshGLDict_.end()) {
+    meshGLDict_[gridName] = Mn::MeshTools::compile(*meshDataDict_[gridName]);
+  } else {
+    meshGLDict_.insert(std::make_pair(
+        gridName, Mn::MeshTools::compile(*meshDataDict_[gridName])));
+  }
+}
+
+void VoxelGrid::generateMesh(const std::string& gridName) {
+  assert(grids_.find(gridName) != grids_.end());
+  std::vector<Mn::UnsignedInt> indices;
+  std::vector<Mn::Vector3> positions;
+  std::vector<Mn::Vector3> normals;
+  std::vector<Mn::Color3> colors;
```

Comment:
Okay let's do some fancy advanced C++ again and save a bunch of overhead that way :) Since there's the same count of positions, normals and colors and MeshData wants them interleaved for the best GPU performance, we could use that data layout right from the start. 

Step 1:

```cpp
// define it outside the function (or maybe in the header?) since addVector...() will need it too
struct VoxelVertex {
  Mn::Vector3 position;
  Mn::Vector3 normal;
  Mn::Vector3 color;
};
Cr::Containers::Array<VoxelVertex> vertices;
```